### DISCUSSION TITLE
Origin of arrows

### DISCUSSION THREAD

**Discussion by AlainZuur** [JenniNiku/gllvm#155] [General]
![Outlier](https://github.com/JenniNiku/gllvm/assets/32171627/a77c431e-2590-4e5c-80c9-3eced465d004)

I've applied an RRR (GLLVM with 2 constrained latent variables).

I (foolishly) ignored a covariate with a very large outlier. When I made an ordiplot with biplot = FALSE, I get a rather odd graph; the origin of the lines/arrows for the covariates/canonical coefficients do not start at the origin. So..I get all these arrows for the canonical coefficients on the left...and the site scores all (but 1) in the right side (yeah...mainly due to a single large outlier in a single covariate).

I've put the ordiplot in ChatGPT, and asked it what is causing this. It came up with this piece of code if biplot=FALSE:

arrows(origin[1], origin[2], ends[, 1] + origin[1], ends[, 2] + origin[2], ...)

where origin <- c(mean(marg[1:2]), mean(marg[3:4]))

So..the origin of the lines is the graph center...which is not necessarily the origin. What is the motivation behind that? Would it not make more sense to let those arrows start in the origin...and use a scaling factor, or a third/fourth axes if scores (Zc) and canonical coefficients are on a different planet?

Kind regards,

Alain

Will deal with the outlier...and that will probably solve the issue....but still...the question remains.

Attached is the ordiplot...and yes...there is the outlier in the lower left...which I only saw after an hour (I thought it was a dirty screen).

**Comment by AlainZuur** [comment]
Trying to answer my own question:

Is that because there is no real reason why those arrows should start at the origin, in the first place? We could as well plot them as labels without arrows?

**Reply by BertvanderVeen** [reply]
Thanks for your question Alain. It has been a while now since I have written this code, but IIRC I did think there is no real sensible point of origin for these arrows.
If you look at the canonical coefficients as a those from a regression of the LV without intercept, there is some case to be made for (0,0) as the point as origin. However, since the LVs are not centered, placing them at that point can also make for odd plots at times, because (0,0) can frequently fall outside of the range the LVs. There is no intercept for the LVs in the model because it is unidentifiable with the species intercepts (in latex: `$\boldsymbol{beta}_0 =\boldsymbol{beta}_0^\star*+\alpha_{lv1}\boldsymbol{theta}_{lv1}+\alpha_{lv2}\boldsymbol{\theta}_{lv2}$`), so that we can in fact consider the model to include LVs-specific intercepts, but the parameterisation does not allow us to retrieve them. Since the LVs are not technically centered, unless the predictors are centered (which should often be done anyways..) placing them at that point can also make for odd plots at times, because (0,0) can frequently fall outside of the range the LVs. 
Ultimately, it is the relative length of the arrows and their angle to the LVs that are of interest, and from that point of view the arrows can be drawn at any point in the plot, and it would not change the conclusion. However, "guessing" the best spot to place them based on the placing of site and species scores is challenging. 
For historical reasons, I chose to draw them from the center of the figure, which depends on the range of the LVs, and consequently also gives awkward positioning at times.

If you have thoughts on better placement, I would be happy to add an argument that gives some flexibility in choosing the origin of the arrows.

**Reply by AlainZuur** [reply]
Hello Bert,

Thanks. Yeah....I thought so.

Actually...it is an example/data set in which I wrongly didn't pay too much attention to data exploration results. Small sample size, 10 covariates, fairly high collinearity, high % of zeros in the species data, and plenty of covariates with a few outliers. I was more interested to see how far you can push GLLVM. I noticed that regularisation, in such a case, can help a lot.

Also...a single outlier in a covariate can be quite influential in this specific data set. I do recall Cajo ter Braak writing somewhere that those canonical coefficients can be 'unstable' in certain scenarios. And is that not the reason that CANOCO is/was using intraset correlations (correlations between CLVs and covariates)? Maybe it would be an idea to add an option to the ordiplot function in which the user can choose between canonical coefficients and intraset correlations? Then you can avoid those weird-looking plots.


Kind regards,

Alain

**Reply by BertvanderVeen** [reply]
I designed the "randomB" option specifically for this purpose (and in no small part motivated by Cajo his suggestions in review), to regularize the canonical coefficients in a random-effects manner. There is something to be said for doing this by default (the models run faster than the fixed-effects equivalent anyway, for technical reasons). The only downside is that I have not implemented a summary table for those coefficients, though it is mentioned in the vignette how to do it yourself.
Out of principle I am against using correlation coefficients that are post-hoc calculated in ordination plots, as they have no connection to the original response variables, no uncertainty quantification, and (generally) there is no good reason (IMO) to do so.

**Reply by AlainZuur** [reply]
Fair enough....but...hypothetical scenario....if you have canonical coefficients that become weird due to collinearity (e.g. a beta is negative due to collinearity whereas its real effect should be positive)...in that case an intraset correlation is useful.

But of course, you could argue that you shouldn't fit such a model in the first instance. 

Anyway...thanks for clarifying the reason why the arrows do not start (always) at the origin.

Alain
